### PR TITLE
GH-2187 keep parent reference in querymodelnode consistent

### PR DESCRIPTION
**Pull request by jeenbroekstra**
GitHub issue resolved: #2187 

Briefly describe the changes proposed in this PR:

- ensure optimizers and other components that manipulate the algebra tree keep the parent node references consistent
- added "cleanup" at end of optimizer pipeline to catch any wrongdoers
- added regression tests for all optimizer implementations

---- 
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [x] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

**Comment by jeenbroekstra** [comment]
From what I can tell most (all?) of the failing tests here are caused specifically by `Var` objects being reused in the path sequence processing. Looks like part of the code is assuming we're using java object identity to do variable unification. That's not right - I think this points to (another) bug in the path sequence processing.

**Comment by jeenbroekstra** [comment]
I will revisit this shortly. We recently did some fairly intensive refactoring of the path expression processing, so there's a chance that we inadvertently addressed this problem as part of that (or if not, at least made it easier to deal with ;))

**Comment by hmottestad** [comment]
I think I remember this being a fairly extensive issue, not limited to paths by any means. Is there a way we could just remove the parent field instead? My confidence with the parent field is really low because it seems much easier to mess it up than to actually get it right.

**Comment by jeenbroekstra** [comment]
> I think I remember this being a fairly extensive issue, not limited to paths by any means. Is there a way we could just remove the parent field instead? My confidence with the parent field is really low because it seems much easier to mess it up than to actually get it right.

We need to be able to somehow refer to a node's parent in order to handle things like replacing nodes (done as part of query optimization/rewriting in several places). We could look into replacing the fixed references with a dynamic treewalker mechanism that starts at the root and then walks down until it finds the first node that has the current node as one of its children, but that would not solve anything I think: we would still run the risk of a node effectively being the child of more than one parent. 

I'd rather we look into fixing the mechanism and putting safeguards against regression in place.

**Comment by jeenbroekstra** [comment]
@hmottestad I have found the immediate cause of the problem, which is a bit of a corner case in that during evaluation of the query a new algebra node is temporarily created (this happens in PathIteration), and that node used uncloned variables. This is the only instance of that kind of behavior that I could find.

I would appreciate it if you could, for the purpose of this PR, remove the bespoke deque you are using `QueryModelTreeToGenericPlanNode` to track these relations (instead of relying on `getParentNode()`). I am not fully following what's happening there, but I'd like to see if we can start relying on the parent node mechanism again for this.

**Comment by jeenbroekstra** [comment]
Btw one idea to "harden" the parent node mechanism is to add a few more assertions (so that we pick up more cases in testing) or even think about putting some exceptions in place.

**Comment by hmottestad** [comment]
@jeenbroekstra I added another failing scenario.

**Comment by hmottestad** [comment]
I feel this issue needs good test coverage before I would trust the parent field.

**Comment by jeenbroekstra** [comment]
Can you please stop adding "temp" tests and other throwaway code? Either put in a proper test or just describe it to me. This is not helping, it's just creating extra work.

**Comment by jeenbroekstra** [comment]
An alternative we can consider is, as you suggested, getting rid of the parentNode field in the algebra model itself, and using a stack of ancestory nodes (very much like you're doing in this specific case) as a general mechanism in the AbstractQueryModelVisitor. That way, the ancestry is always determined dynamically at the point it's needed. Not sure it will solve all corner cases, but it will be easier to test I think.

**Comment by jeenbroekstra** [comment]
From what I can tell, the main problem we're seeing (in all these different cases) is that up to now, using getParentNode  was something  exclusively used _before_ the optimization pipeline kicks in. Several of the visitors in the optimization pipeline mess it about, creating new model nodes without cloning the arguments. The reason this was never a problem before is that we don't really rely on it anywhere else _after_ optimization or evaluation has happened, that's something newly introduced by the explain feature.

I've had a bit of a play with the alternative suggested above, but there are several cases where using that approach is not straightforward at all. We are probably better off keeping `getParentNode()` in place and making sure we come up with some testing strategies that will help us detect "wrongdoers" in the optimization pipeline.

**Comment by hmottestad** [comment]
> Can you please stop adding "temp" tests and other throwaway code? Either put in a proper test or just describe it to me. This is not helping, it's just creating extra work.

I don't know why it is failing. As I've said, when I first created this branch I went digging to try to find the culprit and got lost since there were so many places that seemed to be problematic. The point of the temp test is literally just to show you that the problem was considerably more wide spread.

**Comment by hmottestad** [comment]
My recommendation is to make parent and child fields immutable (final) so that we force the code to always have to decide what the parent and child fields should contain.

**Comment by jeenbroekstra** [comment]
> My recommendation is to make parent and child fields immutable (final) so that we force the code to always have to decide what the parent and child fields should contain.

That would prohibit any sort of optimization or query rewriting.

**Comment by hmottestad** [comment]
> > My recommendation is to make parent and child fields immutable (final) so that we force the code to always have to decide what the parent and child fields should contain.
> 
> That would prohibit any sort of optimization or query rewriting.

No no, we would just make a mechanism that would replace the part of the tree that you want to optimise. Maybe making the child fields immutable isn't necessary, just the parent field.

**Comment by jeenbroekstra** [comment]
Ah I see, if it's _only_ the parent that we make final.... It will still be non-trival though because it means we can't ever move nodes around in the tree. We'll _always_ have to create a copy. We'll also need to adjust the `clone()` method on all node implementations, or rather we'll need to migrate away from using `Object.clone()` an introduce a bespoke `copyOf` method that takes a new parent node as one of its arguments.

**Comment by hmottestad** [comment]
Test wise I was thinking about creating a QueryModelVisitor that uses the stack approach with assertions like the QueryModelTreeToGenericPlanNode, except it would be in the test directory instead. Then we could see how we could integrate it so that it can utilize our current sparql tests. I know we have the W3C sparql tests, maybe we could extend these in the the MemoryStore tests to also test the parent field in the query plan at various levels. Or maybe add a more complex assert to the query execution in general that runs the new QueryModelVisitor to check the parent field against the stack approach.

**Comment by hmottestad** [comment]
Yes, `cloneWithNewParent(....)` was what I had in mind.

**Comment by hmottestad** [comment]
But I have another idea. If it wouldn't degrade performance too much. How about a "cleaner" that can be run after each optimizer to fix up the parent field? Wouldn't solve the issue of nodes being shared instead of cloned, but that problem should only really exist in the parser I hope. This approach would also eliminate any issue from users/developers who add their own optimizers (that could be buggy) and then run into trouble with another optimizer that uses the parent field.

**Comment by jeenbroekstra** [comment]
> But I have another idea. If it wouldn't degrade performance too much. How about a "cleaner" that can be run after each optimizer to fix up the parent field?

Not sure if we should do this after_each_ optimizer, but it's perhaps something we can consider as a "final" optimizer.

**Comment by hmottestad** [comment]
Do you still think this is worth fixing?

**Comment by jeenbroekstra** [comment]
> If one optimizer messes up the parent field and another optimizer uses that field?

There's only so much I'm prepared to do to clean up after optimizers that don't behave properly. I think it's useful to put a failsafe at the end, but the real fix is obviously in the actual optimizers, making sure that they keep the tree consistent. 

I've pushed a draft fix along these lines. 

I suggest that for each individual optimizer we introduce a test case that do, essentially, the same check as I've now added to ParentReferenceCleanerTest. If we make that part of an abstract test suites for optimizer implementations we can cheaply use it as a unit test for each optimizer, making sure that we catch these problems early.

**Comment by hmottestad** [comment]
I feel that one optimizer shouldn't be able to break something as fundamental as the parent field.

**Comment by jeenbroekstra** [comment]
> I feel that one optimizer shouldn't be able to break something as fundamental as the parent field.

Having a testsuite in place that ensures an optimizer implementation behaves is a good step in that direction.

> A good solutions would maybe be to abstract away the code that modifies the query plan tree so that the parent field was always maintained. Though that's not necessarily going to be easy.

I don't think the effort is worth it, to be honest. I'm certainly not intending to spend _that_ much time on this. There's other things far more important to fix/improve.

**Comment by jeenbroekstra** [comment]
I've added some tests for all standard optimizers. Obviously we can extend these test cases to involve different query constructions (filters, functions, aliases, aggregates, etc) but it's a decent start I think.

**Comment by jeenbroekstra** [comment]
This is now ready for final review.

File: core/sail/memory/src/test/java/org/eclipse/rdf4j/sail/memory/QueryPlanRetrievalTest.java [hunk 1]
```diff
@@ -873,4 +873,38 @@ public void testDotTimed() {
 
 	}
 
+	@Test
+	public void testArbitraryLengthPath() {
```
Comment (hmottestad) [inline core/sail/memory/src/test/java/org/eclipse/rdf4j/sail/memory/QueryPlanRetrievalTest.java:877]: I would have been ok with you deleting this test, but it's nice of you to turn it into a functional test.

File: core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/helpers/QueryModelTreeToGenericPlanNode.java [hunk 2]
```diff
@@ -35,10 +37,11 @@ public GenericPlanNode getGenericPlanNode() {
 		return top;
 	}
 
-	// node.getParentNode() is not reliable because nodes are reused and parent is not maintained! This is why we use a
-	// queue to maintain the effective parent stack.
 	@Override
 	protected void meetNode(QueryModelNode node) {
+		assert !visited.containsKey(node);
+		visited.put(node, node);
```
Comment (hmottestad) [inline core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/helpers/QueryModelTreeToGenericPlanNode.java:43]: What was the reason you added this? Did you see any cycles in query plan?

Comment (jeenbroekstra) [reply core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/helpers/QueryModelTreeToGenericPlanNode.java:43]: This is something _you_ added, I think. I assumed it was necessary, but was this part of the throwaway test code as well?

Comment (hmottestad) [reply core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/helpers/QueryModelTreeToGenericPlanNode.java:43]: Yeah. It was.

Comment (hmottestad) [reply core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/helpers/QueryModelTreeToGenericPlanNode.java:43]: I intended for you to just revert my commits when you had figured out how to test the issue properly and how to fix it.

**Comment by hmottestad** [comment]
Looks good to me @jeenbroekstra 

Nice job!

I'm glad you have left my `QueryModelTreeToGenericPlanNode` class as is. Your fixes and tests go a long way, but I'm still nervous about relying on the parent field at the moment. The stack based approach works fine though, so I'm happy to keep it.

Thanks again for your patience with me :)

File: core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/helpers/QueryModelTreeToGenericPlanNode.java (hunk 2 above)
Comment (jeenbroekstra) [reply core/queryalgebra/model/src/main/java/org/eclipse/rdf4j/query/algebra/helpers/QueryModelTreeToGenericPlanNode.java:43]: I'll take it out. See, this is why I don't like throwaway code. It has a nast habit of sticking around.

**Comment by hmottestad** [comment]
When you have removed the temporary cycle detection code, fell free to squash all the commits into one. I've not really contributed any code towards a solution here, so no need to keep my commits around.

File: core/sail/memory/src/test/java/org/eclipse/rdf4j/sail/memory/QueryPlanRetrievalTest.java (hunk 1 above)
Comment (jeenbroekstra) [reply core/sail/memory/src/test/java/org/eclipse/rdf4j/sail/memory/QueryPlanRetrievalTest.java:877]: It's still a useful test to have in place, making sure that the expected explanation is actually produced.

**Comment by jeenbroekstra** [comment]
> When you have removed the temporary cycle detection code, fell free to squash all the commits into one. I've not really contributed any code towards a solution here, so no need to keep my commits around.

I'll stick to the 5 commits now in place :)
